Took the suppressed comment in b1894cb — replying here since suppressed comments have no thread.

You were right that the name overstated what was asserted, and it is the same mistake as the `SubscribeUntil` test earlier in this PR: a test that would have passed against the very thing it exists to catch. Checking `stopped` only proves `Unsubscribe` *asked* the goroutine to stop.

Rather than rename it down to what it checked, I made the stronger claim true. `subscriber` now closes an `exited` channel when `run` returns — a distinct moment from `stopped`, since an in-flight callback still has to finish first — and the test waits on that. The concern behind the test is a goroutine leak, so the goroutine actually returning is the assertion worth having.

Verified by deleting the `stopped` case from the run loop:

```
--- FAIL: TestUnsubscribe_EndsTheDeliveryGoroutine (5.00s)
    ordering_test.go:164: delivery goroutine never returned
```

--- a/events/events.go
+++ b/events/events.go
@@ -218,9 +218,9 @@ func (e *Subscription[T]) Unsubscribe() {
 // own callbacks are serialized.
 func Emit[T Event](evt T) {
 	key := reflect.TypeFor[T]()
-	// Snapshot the callbacks into a slice under the RLock, then drop
+	// Snapshot the subscribers into a slice under the RLock, then drop
 	// the lock before doing anything that could block (the diagnostic
-	// log, the per-callback goroutine spawn). Iterating the underlying
+	// log, the enqueue to each one). Iterating the underlying
 	// map after releasing the lock would race against Unsubscribe's
 	// write lock — `concurrent map iteration and map write` panic
 	// territory under load.

--- a/events/events.go
+++ b/events/events.go
@@ -50,18 +50,10 @@ var (
 // so reaching it means the callback is wedged, not busy.
 const queueDepth = 256
 
-// subscriber owns one subscription's delivery. Each has a single goroutine
-// draining a FIFO queue, which is what makes delivery ordered: Emit used to
-// spawn a fresh goroutine per event per callback, so two Emits raced and the
-// one that landed last was decided by the scheduler.
-//
-// That is not a rare interleaving. Emitting eight events to one subscriber
-// delivered them out of order in 200 of 200 runs, e.g. [0 3 1 2 7 6 5 4] —
-// so the *last* event a consumer saw was routinely not the last one sent.
-// Consumers that render the newest state (the share card renders the phase
-// from whichever frame arrived last) therefore settled on an arbitrary one,
-// which is how a peer that had finished registering kept reading
-// "discovering public IP".
+// subscriber owns one subscription's delivery: a single goroutine draining a
+// FIFO queue, which is what orders it. Emit previously spawned a goroutine per
+// event per callback, leaving the delivery order to the scheduler, so a
+// consumer that renders the newest event received an arbitrary one.
 type subscriber struct {
 	deliver func(any)
 	queue   chan any

--- a/events/ordering_test.go
+++ b/events/ordering_test.go
@@ -193,6 +193,9 @@ func TestUnsubscribe_DropsEventsAlreadyQueued(t *testing.T) {
 				<-release // hold the goroutine inside the first delivery
 			}
 		})
+		subscriptionsMu.RLock()
+		s := subscriptions[reflect.TypeFor[orderedEvt]()][(*Subscription[Event])(sub)]
+		subscriptionsMu.RUnlock()
 
 		// Emit never blocks, so all four are queued before the callback for
 		// the first one has finished.
@@ -203,7 +206,14 @@ func TestUnsubscribe_DropsEventsAlreadyQueued(t *testing.T) {
 		sub.Unsubscribe()
 		close(release)
 
-		time.Sleep(20 * time.Millisecond)
+		// Waiting for the goroutine to return is exact, where a sleep would
+		// only be probably-long-enough — and it cannot pass by finishing
+		// before a late delivery that was still coming.
+		select {
+		case <-s.exited:
+		case <-time.After(5 * time.Second):
+			t.Fatalf("round %d: delivery goroutine never returned", r)
+		}
 		mu.Lock()
 		delivered := append([]int(nil), got...)
 		mu.Unlock()

--- a/ipc/client_events_nonmobile.go
+++ b/ipc/client_events_nonmobile.go
@@ -59,9 +59,8 @@ func (c *Client) DataCapStream(ctx context.Context, handler func(account.DataCap
 // → registering → verifying → serving on Start, stopping → idle on Stop,
 // error on failure). Each frame is a peer.StatusEvent JSON whose .Status
 // is the live snapshot at the moment the event fired — consumers SHOULD
-// re-render on every frame rather than diffing, since events.Emit's
-// per-callback goroutine can land Start phases out of order. Blocks until
-// ctx is cancelled.
+// re-render on every frame rather than diffing. Frames arrive in order, so
+// the last one seen is the current phase. Blocks until ctx is cancelled.
 func (c *Client) PeerStatusEvents(ctx context.Context, handler func(peer.StatusEvent)) error {
 	return c.sseRetryLoop(ctx, peerStatusEventsEndpoint, func(data []byte) {
 		var evt peer.StatusEvent

--- a/ipc/server.go
+++ b/ipc/server.go
@@ -496,10 +496,10 @@ func (s *localapi) peerStatusHandler(w http.ResponseWriter, r *http.Request) {
 
 // peerStatusEventsHandler streams peer.StatusEvent over SSE. The wire
 // payload is always the live snapshot from PeerStatus(), not the event's
-// captured value: events.Emit dispatches each callback on its own goroutine
-// so a quick start→stop pair can land out of order, and we'd send a stale
-// "active" after a "inactive". Reading the live snapshot when the trigger
-// fires guarantees the SSE consumer's last message reflects current state.
+// captured value: the trigger below coalesces, so a burst of phases becomes
+// one wakeup and the captured value could be one the consumer never needed.
+// Reading the live snapshot when the trigger fires guarantees the SSE
+// consumer's last message reflects current state.
 func (s *localapi) peerStatusEventsHandler(w http.ResponseWriter, r *http.Request) {
 	flusher := sseWriter(w)
 	if flusher == nil {
@@ -538,9 +538,7 @@ func (s *localapi) peerStatusEventsHandler(w http.ResponseWriter, r *http.Reques
 // each accept/close on the local samizdat-in. Unlike peerStatusEventsHandler
 // (which always sends the live snapshot), each emit's captured value is
 // what the consumer needs here — the Source IP and +1/-1 state ARE the
-// payload, not a periodic poll. Out-of-order +1/-1 from events.Emit's
-// per-callback goroutine is fine: the consumer (lantern-core's globe-arc
-// renderer) keys arcs by source, so it handles re-orderings naturally.
+// payload, not a periodic poll, so none may be collapsed into a wakeup.
 //
 // The events package lives in this process (lanternd); cross-process
 // consumers in Liblantern can only receive these via this SSE stream,

--- a/ipc/server.go
+++ b/ipc/server.go
@@ -562,7 +562,8 @@ func (s *localapi) peerConnectionEventsHandler(w http.ResponseWriter, r *http.Re
 		case queue <- evt:
 		default:
 			// queue full — drop. SSE consumer is too slow; better to
-			// lose this event than to back up the events.Emit goroutine.
+			// lose this event here than to stall this subscription's
+			// delivery goroutine and have the bus drop from further back.
 		}
 	})
 	defer sub.Unsubscribe()

--- a/peer/peer_test.go
+++ b/peer/peer_test.go
@@ -880,10 +880,11 @@ func TestClient_StatusEventEmittedOnStartAndStop(t *testing.T) {
 }
 
 // drainPhases reads up to n StatusEvents from got and returns them
-// keyed by Phase (last event per phase wins). Used by tests that need
-// set-membership semantics rather than strict ordering because
-// events.Emit's per-callback goroutines deliver out of order under
-// the runtime's scheduling.
+// keyed by Phase (last event per phase wins). Used by tests that care
+// which phases occurred rather than in what order. The bus does deliver
+// them in order now, so those tests could assert the sequence instead;
+// they do not, because the phase a given step emits is not what they are
+// pinning.
 func drainPhases(t *testing.T, got <-chan StatusEvent, n int) map[Phase]StatusEvent {
 	t.Helper()
 	out := make(map[Phase]StatusEvent, n)

--- a/unbounded/unbounded_test.go
+++ b/unbounded/unbounded_test.go
@@ -664,8 +664,8 @@ func TestConnectionEventBridge(t *testing.T) {
 	cb := *captured.Load()
 	require.NotNil(t, cb, "OnConnectionChangeFunc must be installed on bfOpt")
 
-	// Buffered enough that events.Emit's per-callback goroutines
-	// can deposit before the test reads.
+	// Buffered so the subscription's delivery goroutine can deposit
+	// without blocking before the test reads.
 	ch := make(chan ConnectionEvent, 4)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()

